File: data/readable_format/cpp/file536.cpp
#include <iostream>
#include <cassert>

// Base class corresponding to Fortran 'type foo'
class foo {
public:
    // Virtual function to allow overriding in derived class
    virtual void bar(int& j) {
        // Base class implementation
        std::cout << "Base bar, j: " << j << std::endl;
    }
    
    // Virtual destructor for proper cleanup of derived classes
    virtual ~foo() {}
};

// Derived class corresponding to Fortran 'type, extends(foo) :: extfoo'
class extfoo : public foo {
public:
    // Override the bar function
    void bar(int& j) override {
        // Derived class implementation
        j += 1; // Example modification to show intent(inout)
        std::cout << "Derived bar, j: " << j << std::endl;
    }
};

// Simple test function to verify functionality
void test() {
    int j_base = 1;
    int j_derived = 1;
    
    foo* basePtr = new foo();
    basePtr->bar(j_base);
    assert(j_base == 1); // Assuming base.bar doesn't change j, it should still be 1
    
    foo* derivedPtr = new extfoo();
    derivedPtr->bar(j_derived);
    assert(j_derived == 2); // Assuming derived.bar increments j, it should now be 2
    
    delete basePtr;
    delete derivedPtr;
    
    std::cout << "All tests passed!" << std::endl;
}

int main() {
    test();
    
    return 0;
}

File: data/readable_format/fortran/file536.f90
module foo_mod
  implicit none

  type, public :: foo
  contains
    procedure :: bar => base_bar
  end type foo

  type, extends(foo), public :: extfoo
  contains
    procedure :: bar => derived_bar
  end type extfoo

contains

  subroutine base_bar(this, j)
    class(foo), intent(inout) :: this
    integer, intent(in) :: j
    print *, "In base_bar with j=", j
  end subroutine base_bar

  subroutine derived_bar(this, j)
    class(extfoo), intent(inout) :: this
    integer, intent(in) :: j
    print *, "In derived_bar with j=", j
  end subroutine derived_bar

end module foo_mod

program test_foo_extfoo
  use foo_mod
  implicit none

  type(foo) :: aFoo
  type(extfoo) :: aExtFoo
  integer :: j

  ! Testing the base class method
  j = 10
  call aFoo%bar(j)

  ! Testing the derived class method
  j = 20
  call aExtFoo%bar(j)
end program test_foo_extfoo